Sources/App/Features/Presentations/Controllers/PresentationViewController.swift

```diff
@@ -0,0 +1,51 @@
+//
+//  PresentationController.swift
+//  
+//
+//  Created by Joe Williams on 07/06/2022.
+//
+
+import Vapor
+import Fluent
+
+struct PresentationViewController: RouteCollection {
+
+    private struct PresentationContext: Content {
+        let presentation: Presentation?
+        let speakers: [Speaker]
+        let events: [Event]
+    }
+    
+    func boot(routes: RoutesBuilder) throws {
+        routes.get("", use: onCreate)
+        routes.get(":id", use: onEdit)
+    }
+        
+    private func onCreate(request: Request) async throws -> View {
+        guard request.user?.role == .admin else {
```

Comment:
An extension would be really great here, so we can do `request.hasAdminPrivileges()`.

Even better if it was middleware for a group of routes so this code didnt even have to leak out into the route handler

Reply:
I don't know if it's possible now in Vapor 4, but I remember wrapping some existing middleware around routes in Vapor 3 that allowed you to guard routes based on user permissions. Might be the case of a custom middleware here through which is still definitely worth doing

Reply:
Middleware is still available in Vapor 4 and grouping by it is also still there, so could be a nice improvement but would probably do that as another ticket

Reply:
An extension already exists I think, pretty sure I added one along the way somewhere